DUI3-DX/Connectors/ArcGIS/Speckle.Connectors.ArcGIS3/Utils/ArcGisDocumentStore.cs

```diff
@@ -9,22 +14,81 @@ public class ArcGISDocumentStore : DocumentModelStore
   public ArcGISDocumentStore(JsonSerializerSettings serializerOption)
     : base(serializerOption)
   {
-    // POC: Subscribe here document related events like OnSave, OnClose, OnOpen etc...
     ActiveMapViewChangedEvent.Subscribe(OnMapViewChanged);
+    ProjectSavingEvent.Subscribe(OnProjectSaving);
+    ProjectClosingEvent.Subscribe(OnProjectClosing);
   }
 
+  private Task OnProjectClosing(ProjectClosingEventArgs arg)
+  {
+    if (MapView.Active is null)
+    {
+      return Task.CompletedTask;
+    }
+
+    WriteToFile();
+    return Task.CompletedTask;
+  }
+
+  private Task OnProjectSaving(ProjectEventArgs arg)
+  {
+    if (MapView.Active is null)
+    {
+      return Task.CompletedTask;
+    }
+
+    WriteToFile();
+    return Task.CompletedTask;
+  }
+
+  /// <summary>
+  /// On map view switch, this event trigger twice, first for outgoing view, second for incoming view.
+  /// </summary>
   private void OnMapViewChanged(ActiveMapViewChangedEventArgs args)
   {
-    OnDocumentChanged();
+    if (args.OutgoingView is not null)
+    {
+      WriteToFileWithMap(args.OutgoingView.Map);
+    }
+
+    if (args.IncomingView is not null)
+    {
+      IsDocumentInit = true;
+      ReadFromFile();
+      OnDocumentChanged();
+    }
   }
 
-  public override void WriteToFile()
+  private void WriteToFileWithMap(Map map)
   {
-    // Implement the logic to save it to file
+    string serializedModels = Serialize();
+
+    XDocument xmlDocument = new(new XElement("metadata", new XElement("SpeckleModelCards", serializedModels)));
+
+    QueuedTask.Run(() =>
+    {
+      map.SetMetadata(xmlDocument.ToString());
```

Comment:
I have noticed this is overwriting existing file metadata that might look like this: 
![image](https://github.com/specklesystems/speckle-sharp/assets/89912278/16c25ef7-75fc-49ee-af75-167e5cea7fa5)

Maybe we can append new tag to the existing metadata?

Reply:
Good point, let me fix it!